### PR TITLE
P4-OVS: Add support for hotplugging vhost user ports to qemu based VM

### PR DESCRIPTION
**Pull request by nupuruttarwar**
This patch adds support for hotplugging vhost user ports for qemu
based VM

Title: Add support for hotplugging vhost user ports to qemu based VM
Change-type: FeatureImplementation
Signed-off-by: nupuruttarwar <nupur.uttarwar@intel.com>

### CODE REVIEW

**Comment by stolsma** [comment]
Great development but for me it doesn't feel architecturally right to add this code to ovs. I would expect this code in a (ipdk) orchestration module/controller/library because if the vhost port is created on a physical target qemu/kvm could even be living on a seperate cpu. For example when ovs is running on a dpu/ipu and qemu/kvm on the host cpu (could work if the qemu socket ip is reachable from ipu/dpu, I know).

Another thing that I was thinking about is why this function (adding a vhost port to qemu) isn't already implemented in the current ovs code? Wouldn't be surprised if this will give some backpressure from the ovs maintainers.

**Comment by nupuruttarwar** [comment]
> Great development but for me it doesn't feel architecturally right to add this code to ovs. I would expect this code in a (ipdk) orchestration module/controller/library because if the vhost port is created on a physical target qemu/kvm could even be living on a seperate cpu. For example when ovs is running on a dpu/ipu and qemu/kvm on the host cpu (could work if the qemu socket ip is reachable from ipu/dpu, I know).
> 
> Another thing that I was thinking about is why this function (adding a vhost port to qemu) isn't already implemented in the current ovs code? Wouldn't be surprised if this will give some backpressure from the ovs maintainers.

Hi Sanders,
Thanks for the review. vhost port is created using Openconfig and currently P4-OVS has the interface to Openconfig, hence we need to have the hotplug changes here as well. I agree with your comment on P4-OVS running on IPU vs host and for that we would want part of this code (where it hotplugs the port) to be able to differentiate between the 2 scenarios. We are discussing the baremetal  use case as well and a possible solution to have this code in the target libraries instead of P4-OVS. We can make those changes when the final decision has been made on IPU side. For P4-DPDK target as well, we could move the vhost hotplug to the p4-dpdk-target library once we find a common place to put it.

File: external/PATCH-01-STRATUM [hunk 1]
```diff
@@ -694,9 +694,22 @@ index f553054d..3f41e9e1 100644
  }
 
 diff --git a/stratum/hal/lib/common/common.proto b/stratum/hal/lib/common/common.proto
-index f7a72d70..45736661 100644
+index f7a72d70..7d5f2fd6 100644
 --- a/stratum/hal/lib/common/common.proto
 +++ b/stratum/hal/lib/common/common.proto
+@@ -46,6 +46,12 @@ enum Platform {
```
Comment (Namrata-intel) [inline external/PATCH-01-STRATUM]: why is this needed? Did you add this?

Comment (nupuruttarwar) [reply external/PATCH-01-STRATUM]: This is just a reference for patch file where to add new enum structure SWBackendQemuHotplugStatus. Only the lines with '++' is new code. Others are just reference line used when applying the patch

File: p4proto/bfIntf/bf_chassis_manager.cc [hunk 2]
```diff
@@ -169,6 +281,55 @@ ::util::Status BfChassisManager::ValidateAndAdd(uint64 node_id, uint32 port_id,
       LOG(INFO) << "SDK_PORT ID while validating = " << sdk_port_id;
       RETURN_IF_ERROR(AddPortHelper(node_id, unit, sdk_port_id, singleton_port, &config));
   }
+
+  if((validate & GNMI_CONFIG_HOTPLUG) == GNMI_CONFIG_HOTPLUG) {
```
Comment (Namrata-intel) [inline p4proto/bfIntf/bf_chassis_manager.cc]: Please add space after if ((

File: p4proto/bfIntf/bf_chassis_manager.cc [hunk 3]
```diff
@@ -169,6 +281,55 @@ ::util::Status BfChassisManager::ValidateAndAdd(uint64 node_id, uint32 port_id,
...
+      std::string chardev_id = config.qemu_vm_chardev_id.c_str();
+      std::string cmd;
+      uint64 mac_address = config.qemu_vm_mac_address;
+      std::string string_mac = (absl::StrFormat("%02x:%02x:%02x:%02x:%02x:%02x", (mac_address >> 40) & 0xFF,
+                                                   (mac_address >> 32) & 0xFF, (mac_address >> 24) & 0xFF,
+                                                   (mac_address >> 16) & 0xFF, (mac_address >> 8) & 0xFF,
+                                                    mac_address & 0xFF));
+
+      if((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
+       {
```
Comment (Namrata-intel) [inline p4proto/bfIntf/bf_chassis_manager.cc]: please add space after if (( and bring { in the same line as if

**Comment by nupuruttarwar** [comment]
Corresponding commit in p4_dpdk_target_repo - https://github.com/p4lang/p4-dpdk-target/pull/24

**Comment by nupuruttarwar** [comment]
Closing the Github PR since this repo doesn't use GitHub pull requests to manage code contributions anymore and has migrated to Gerrit. Hotplug feature will be tracked via https://review.gerritipdk.com/gerrit/c/ipdk-io/ovs/+/241
